src/execution/execute.ts

```diff
@@ -210,12 +268,58 @@ export interface ExecutionArgs {
  *
  * If the arguments to this function do not result in a legal execution context,
  * a GraphQLError will be thrown immediately explaining the invalid input.
+ *
+ * This function does not support incremental delivery (`@defer` and `@stream`).
+ * If an operation which would defer or stream data is executed with this function,
+ * it will throw or reject instead. Use `experimentalExecuteIncrementally` if you
+ * want to support incremental delivery.
  */
-export function execute(
+export function execute(args: ExecutionArgs): PromiseOrValue<ExecutionResult> {
+  const result = executeSyncOrIncrementally(args);
+  if (!isPromise(result)) {
+    return result;
+  }
+
+  return result.then((incrementalResult) => {
+    if ('singleResult' in incrementalResult) {
+      return incrementalResult.singleResult;
+    }
+    throw new GraphQLError(
+      'Executing this GraphQL operation would unexpectedly produce multiple payloads (due to @defer or @stream directive)',
+    );
+  });
+}
+
+/**
+ * Implements the "Executing requests" section of the GraphQL specification,
+ * including `@defer` and `@stream` as proposed in
+ * https://github.com/graphql/graphql-spec/pull/742
+ *
+ * This function returns a Promise of an ExperimentalExecuteIncrementallyResults
+ * object. This object either contains a single ExecutionResult as
+ * `singleResult`, or an `initialResult` and a stream of `subsequentResults`.
+ *
+ * If the arguments to this function do not result in a legal execution context,
+ * a GraphQLError will be thrown immediately explaining the invalid input.
+ */
+export function experimentalExecuteIncrementally(
   args: ExecutionArgs,
-): PromiseOrValue<
-  ExecutionResult | AsyncGenerator<AsyncExecutionResult, void, void>
-> {
+): Promise<ExperimentalExecuteIncrementallyResults> {
+  const result = executeSyncOrIncrementally(args);
+  if (isPromise(result)) {
+    return result;
+  }
+  // Always return a Promise for a consistent API.
+  return Promise.resolve({ singleResult: result });
+}
+
+/**
+ * Helper for the other execute functions. Returns an ExecutionResult
+ * synchronously if all resolvers return non-Promises
+ */
+function executeSyncOrIncrementally(
```

Comment:
Do we need this combination?

Reply:
I agree, it with @yaacovCR on that. It's DRY and saves a few lines. But given all the planned changes to `execute` API, it would be better to just duplicate those lines in both functions.

Reply:
@yaacovCR Also, it's a non-exported function, just a utility one to be DRY.
But I still think it's somewhat confusing and don't worth the DRY benefits it gives.